src/components/patternfly/HeatMap.js

```diff
@@ -1,115 +1,117 @@
-import React from 'react'
+import React, { useEffect, useRef, useState } from 'react'
 import PropTypes from 'prop-types'
-import $ from 'jquery'
 import d3 from 'd3'
+import { Tooltip } from '@patternfly/react-core'
 
 // PatternFly reference:
-//  http://www.patternfly.org/pattern-library/data-visualization/heat-map/
-
-// TODO(vs) rewrite in pure React & consider contributing to patternfly-react
-// TODO: PF4 tooltip
-
-class HeatMap extends React.Component {
-  componentDidMount () {
-    this._generateHeatMap(this.props)
+//  https://www.patternfly.org/v3/pattern-library/data-visualization/heat-map/
+
+function determineBlockSize ({ containerWidth, containerHeight, numberOfBlocks, maxBlockSize }) {
+  const x = containerWidth
+  const y = containerHeight
+  const n = numberOfBlocks
+  const px = Math.ceil(Math.sqrt(n * x / y))
+  const py = Math.ceil(Math.sqrt(n * y / x))
+
+  let sx
+  if (Math.floor(px * y / x) * px < n) {
+    sx = y / Math.ceil(px * y / x)
+  } else {
+    sx = x / px
   }
 
-  componentDidUpdate (newProps) {
-    this._updateHeatMap(newProps)
+  let sy
+  if (Math.floor(py * x / y) * py < n) {
+    sy = x / Math.ceil(x * py / y)
+  } else {
+    sy = y / py
   }
 
-  render () {
-    return (
-      <div className='heatmap-container' style={this.props.containerStyle} ref={e => { this._heatMapContainer = e }}>
-        <svg className='heatmap-svg' />
-      </div>
-    )
-  }
+  sx = Math.min(sx, maxBlockSize)
+  sy = Math.min(sy, maxBlockSize)
 
-  _generateHeatMap ({ data, thresholds, maxBlockSize, blockPadding, onBlockClick }) {
-    const containerWidth = this._heatMapContainer.clientWidth
-    const containerHeight = this._heatMapContainer.clientHeight
+  return Math.max(sx, sy)
+}
+
+export const DEFAULT_THRESHOLD_COLORS = ['#D4F0FA', '#F9D67A', '#EC7A08', '#CE0000']
 
-    const blockSize = this._determineBlockSize({
+const HeatMap = ({
+  id,
+  data,
+  thresholds = {
+    domain: [0.7, 0.8, 0.9],
+    colors: DEFAULT_THRESHOLD_COLORS,
+  },
+  maxBlockSize = 50,
+  blockPadding = 1,
+  containerStyle = {
+    height: 100,
+  },
+  onBlockClick = () => {},
+}) => {
+  const containerRef = useRef()
+  const [{ blockSize, numberOfRows }, setRenderParams] = useState({})
+  const [allBlocksActive, setAllBlocksActive] = useState(true)
+  const [mouseOverBlock, setMouseOverBlock] = useState(-1)
+
+  useEffect(() => {
```

Comment:
why we need `useEffect` hook? if this is only for performance then perhaps we could use `useMemo`?

Reply:
The effect uses the `containerRef` and needs to run AFTER the initial render is done and the DOM is all updated.  The ref won't exist until then.

From https://reactjs.org/docs/hooks-effect.html:

> What does useEffect do? By using this Hook, you tell React that your component needs to do something after render. React will remember the function you passed (we’ll refer to it as our “effect”), and call it later after performing the DOM updates. In this effect, we set the document title, but we could also perform data fetching or call some other imperative API.

The important lines that need the ref:
```jsx
  useEffect(() => {
    const containerWidth = containerRef.current.clientWidth
    const containerHeight = containerRef.current.clientHeight
```

Reply:
Ok. Tha makes sense. However combining element size detection with block calculations makes this algorithm look complex. How about decoupling them? i.e. similar as [useElementSize](https://usehooks-ts.com/react-hook/use-element-size) hook. We could also consider re-using the hook (or similar).

Reply:
@sjd78 I had some visual problems after resizing the window. Not sure if this is due to this part but the values are calculated after resize (the hook depends only on `data` and  `maxBlockSize`).

Reply:
Nice hook, I'll use that.